File: internal/menu/menu.go
// Package menu renders the shared MARS management menu used by both the
// relay and agent binaries when the user runs the `ms` subcommand.
package menu

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strings"

	"mars/internal/service"
)

// Spec describes one end of MARS for the menu.
type Spec struct {
	Title        string // shown at the top, e.g. "MARS 中转（relay）"
	ServiceName  string // e.g. "mars-relay"
	ConfigPath   string // absolute path to the YAML on disk
	Binary       string // absolute path to the running binary (for install)
	InstallSpec  service.Spec
	EditConfig   func() error   // called when user picks "修改配置"
	RunForeground func() error  // start the service in foreground (current process)
}

// Run shows the menu loop. Returns when the user picks "退出".
func Run(sp Spec) error {
	in := bufio.NewReader(os.Stdin)
	for {
		st, _ := service.QueryStatus(sp.ServiceName)
		renderHeader(sp, st)

		var options []option
		if st.Installed {
			options = installedOptions(sp)
		} else {
			options = notInstalledOptions(sp)
		}

		for _, o := range options {
			fmt.Printf(" %s) %s\n", o.key, o.label)
		}
		fmt.Println(" q) 退出")
		fmt.Println("-----------------------------------------------------")
		fmt.Print("选择： ")

		line, err := in.ReadString('\n')
		if err != nil {
			return nil
		}
		choice := strings.TrimSpace(line)
		if choice == "q" || choice == "Q" || choice == "" {
			return nil
		}

		handled := false
		for _, o := range options {
			if o.key == choice {
				if err := o.action(); err != nil {
					fmt.Printf("\n!! 操作失败：%v\n", err)
				} else if o.successMsg != "" {
					fmt.Printf("\n✓ %s\n", o.successMsg)
				}
				handled = true
				break
			}
		}
		if !handled {
			fmt.Printf("\n!! 无效选项：%q\n", choice)
		}
		fmt.Println()
		fmt.Print("按回车继续 ...")
		_, _ = in.ReadString('\n')
	}
}

type option struct {
	key        string
	label      string
	action     func() error
	successMsg string
}

func renderHeader(sp Spec, st service.Status) {
	clearScreen()
	fmt.Println("=====================================================")
	fmt.Printf(" %s\n", sp.Title)
	fmt.Println("=====================================================")
	fmt.Printf(" 服务名  ： %s\n", sp.ServiceName)
	fmt.Printf(" 配置文件： %s\n", sp.ConfigPath)
	fmt.Printf(" 状态    ： %s\n", renderStatus(st))
	fmt.Println("-----------------------------------------------------")
}

func renderStatus(st service.Status) string {
	if !st.Installed {
		return "未安装为系统服务"
	}
	parts := []string{}
	if st.Running {
		parts = append(parts, "运行中")
	} else {
		parts = append(parts, "已停止")
	}
	if st.Enabled {
		parts = append(parts, "开机自启")
	} else {
		parts = append(parts, "未设为自启")
	}
	label := strings.Join(parts, " / ")
	if st.Detail != "" {
		label += "  (" + st.Detail + ")"
	}
	return label
}

func installedOptions(sp Spec) []option {
	return []option{
		{
			key:    "1",
			label:  "查看当前配置",
			action: func() error { return viewConfig(sp.ConfigPath) },
		},
		{
			key:        "2",
			label:      "修改配置（保存后需重启服务生效）",
			action:     sp.EditConfig,
			successMsg: "配置已保存",
		},
		{
			key:        "3",
			label:      "启动服务",
			action:     func() error { return service.Start(sp.ServiceName) },
			successMsg: "已发送启动指令",
		},
		{
			key:        "4",
			label:      "停止服务",
			action:     func() error { return service.Stop(sp.ServiceName) },
			successMsg: "已停止",
		},
		{
			key:        "5",
			label:      "重启服务",
			action:     func() error { return service.Restart(sp.ServiceName) },
			successMsg: "已重启",
		},
		{
			key:        "6",
			label:      "设为开机自启",
			action:     func() error { return service.Enable(sp.ServiceName) },
			successMsg: "已设为开机自启",
		},
		{
			key:        "7",
			label:      "取消开机自启",
			action:     func() error { return service.Disable(sp.ServiceName) },
			successMsg: "已取消开机自启",
		},
		{
			key:        "8",
			label:      "卸载服务",
			action:     func() error { return service.Uninstall(sp.ServiceName) },
			successMsg: "已卸载",
		},
	}
}

func notInstalledOptions(sp Spec) []option {
	opts := []option{
		{
			key:    "1",
			label:  "查看当前配置",
			action: func() error { return viewConfig(sp.ConfigPath) },
		},
		{
			key:        "2",
			label:      "修改配置",
			action:     sp.EditConfig,
			successMsg: "配置已保存",
		},
		{
			key:        "3",
			label:      "注册为系统服务并启动",
			action:     func() error { return service.Install(sp.InstallSpec) },
			successMsg: "已注册并启动",
		},
	}
	if sp.RunForeground != nil {
		opts = append(opts, option{
			key:    "4",
			label:  "前台运行一次（Ctrl+C 停止）",
			action: sp.RunForeground,
		})
	}
	return opts
}

// tokenLine matches a YAML line like `token: abcdef...` so we can redact the
// long secret when showing the config on screen.
var tokenLine = regexp.MustCompile(`(?m)^(\s*token\s*:\s*)(\S+)`)

func viewConfig(path string) error {
	if path == "" {
		return fmt.Errorf("配置路径为空")
	}
	b, err := os.ReadFile(path)
	if err != nil {
		return err
	}
	redacted := tokenLine.ReplaceAllStringFunc(string(b), func(m string) string {
		parts := tokenLine.FindStringSubmatch(m)
		prefix, val := parts[1], parts[2]
		if len(val) <= 12 {
			return prefix + val
		}
		return prefix + val[:8] + "...（已省略）"
	})
	fmt.Println()
	fmt.Println("-----------------------------------------------------")
	fmt.Printf(" 配置文件：%s\n", path)
	fmt.Println("-----------------------------------------------------")
	fmt.Print(redacted)
	if !strings.HasSuffix(redacted, "\n") {
		fmt.Println()
	}
	fmt.Println("-----------------------------------------------------")
	return nil
}

func clearScreen() {
	// ANSI: clear + home cursor. Works in Windows Terminal, cmd.exe on Win10+,
	// all modern *nix terminals. Worst case you see the escape once — harmless.
	fmt.Print("\x1b[2J\x1b[H")
}

File: rust/src/menu.rs
use std::io::Write;
use std::pin::Pin;

use anyhow::{Context, Result};

use crate::service;

pub type MenuFuture = Pin<Box<dyn std::future::Future<Output = Result<()>>>>;

pub struct Spec {
    pub title: String,
    pub service_name: String,
    pub config_path: String,
    pub install_spec: service::Spec,
    pub edit_config: Box<dyn Fn() -> MenuFuture>,
    pub run_foreground: Option<Box<dyn Fn() -> MenuFuture>>,
}

pub fn redact_config_text(input: &str) -> String {
    input
        .lines()
        .map(|line| {
            let trimmed = line.trim_start();
            if let Some(value) = trimmed.strip_prefix("token:") {
                let prefix_len = line.len() - trimmed.len();
                let prefix = &line[..prefix_len];
                let value = value.trim();
                if value.len() > 12 {
                    format!("{prefix}token: {}...（已省略）", &value[..8])
                } else {
                    line.to_string()
                }
            } else {
                line.to_string()
            }
        })
        .collect::<Vec<_>>()
        .join("\n")
}

pub async fn run(spec: Spec) -> Result<()> {
    loop {
        let status = service::query_status(&spec.service_name).unwrap_or_default();
        render_header(&spec, &status);
        let options = if status.installed {
            installed_options(spec.run_foreground.is_some())
        } else {
            not_installed_options(spec.run_foreground.is_some())
        };
        for (key, label) in &options {
            println!(" {key}) {label}");
        }
        println!(" q) 退出");
        println!("-----------------------------------------------------");
        print!("选择： ");
        std::io::stdout().flush().context("flush stdout")?;

        let mut line = String::new();
        std::io::stdin()
            .read_line(&mut line)
            .context("read stdin")?;
        let choice = line.trim();
        if choice.eq_ignore_ascii_case("q") || choice.is_empty() {
            return Ok(());
        }

        let result = match choice {
            "1" => view_config(&spec.config_path),
            "2" => (spec.edit_config)().await,
            "3" if status.installed => service::start(&spec.service_name),
            "3" => service::install(spec.install_spec.clone()),
            "4" if status.installed => service::stop(&spec.service_name),
            "4" => {
                if let Some(run_foreground) = &spec.run_foreground {
                    run_foreground().await
                } else {
                    Ok(())
                }
            }
            "5" if status.installed => service::restart(&spec.service_name),
            "6" if status.installed => service::enable(&spec.service_name),
            "7" if status.installed => service::disable(&spec.service_name),
            "8" if status.installed => service::uninstall(&spec.service_name),
            _ => {
                println!("\n!! 无效选项：{choice:?}");
                Ok(())
            }
        };

        if let Err(err) = result {
            println!("\n!! 操作失败：{err:#}");
        }
        println!();
        print!("按回车继续 ...");
        let _ = std::io::stdout().flush();
        let mut pause = String::new();
        let _ = std::io::stdin().read_line(&mut pause);
    }
}

fn installed_options(_has_foreground: bool) -> Vec<(&'static str, &'static str)> {
    vec![
        ("1", "查看当前配置"),
        ("2", "修改配置（保存后需重启服务生效）"),
        ("3", "启动服务"),
        ("4", "停止服务"),
        ("5", "重启服务"),
        ("6", "设为开机自启"),
        ("7", "取消开机自启"),
        ("8", "卸载服务"),
    ]
}

fn not_installed_options(has_foreground: bool) -> Vec<(&'static str, &'static str)> {
    let mut options = vec![
        ("1", "查看当前配置"),
        ("2", "修改配置"),
        ("3", "注册为系统服务并启动"),
    ];
    if has_foreground {
        options.push(("4", "前台运行一次（Ctrl+C 停止）"));
    }
    options
}

fn render_header(spec: &Spec, status: &service::Status) {
    print!("\x1b[2J\x1b[H");
    println!("=====================================================");
    println!(" {}", spec.title);
    println!("=====================================================");
    println!(" 服务名  ： {}", spec.service_name);
    println!(" 配置文件： {}", spec.config_path);
    println!(" 状态    ： {}", render_status(status));
    println!("-----------------------------------------------------");
}

fn render_status(status: &service::Status) -> String {
    if !status.installed {
        return "未安装为系统服务".to_string();
    }
    let mut parts = Vec::new();
    parts.push(if status.running {
        "运行中"
    } else {
        "已停止"
    });
    parts.push(if status.enabled {
        "开机自启"
    } else {
        "未设为自启"
    });
    let mut label = parts.join(" / ");
    if !status.detail.is_empty() {
        label.push_str("  (");
        label.push_str(&status.detail);
        label.push(')');
    }
    label
}

fn view_config(path: &str) -> Result<()> {
    let content = std::fs::read_to_string(path).with_context(|| format!("read {path}"))?;
    println!();
    println!("-----------------------------------------------------");
    println!(" 配置文件：{path}");
    println!("-----------------------------------------------------");
    println!("{}", redact_config_text(&content));
    println!("-----------------------------------------------------");
    Ok(())
}
